### DISCUSSION TITLE
failis.. Failed to grab device

### DISCUSSION THREAD

**Discussion by xpusostomos** [xremap/xremap#603] [Q&A]
multiple times per day, xremap stops working and needs restarting. I don't know if the below error is related to it, but its in the logs periodically so maybe it's related to this.

```
Nov 09 17:17:25 P14s-Arch keymap[442156]: Failed to grab device 'USB-HID Keyboard' at '/dev/input/event19' due to: No such device (os error 19)
```

and some more log context:
```
Nov 09 17:14:09 P14s-Arch keymap[442156]: /dev/input/event7 : HD-Audio Generic HDMI/DP,pcm=3
Nov 09 17:14:09 P14s-Arch keymap[442156]: /dev/input/event8 : HD-Audio Generic HDMI/DP,pcm=7
Nov 09 17:14:09 P14s-Arch keymap[442156]: /dev/input/event9 : HD-Audio Generic HDMI/DP,pcm=8
Nov 09 17:14:09 P14s-Arch keymap[442156]: ------------------------------------------------------------------------------
Nov 09 17:14:09 P14s-Arch keymap[442156]: Selected keyboards automatically since --device options weren't specified:
Nov 09 17:14:09 P14s-Arch keymap[442156]: ------------------------------------------------------------------------------
Nov 09 17:14:09 P14s-Arch keymap[442156]: /dev/input/event15: USB-HID Keyboard
Nov 09 17:14:09 P14s-Arch keymap[442156]: /dev/input/event19: USB-HID Keyboard
Nov 09 17:14:09 P14s-Arch keymap[442156]: /dev/input/event4 : AT Translated Set 2 keyboard
Nov 09 17:14:09 P14s-Arch keymap[442156]: ------------------------------------------------------------------------------
Nov 09 17:17:24 P14s-Arch keymap[442156]: Found a removed device. Reselecting devices.
Nov 09 17:17:24 P14s-Arch keymap[442156]: Failed to ungrab device 'USB-HID Keyboard' at '/dev/input/event15' due to: No such device (os error 19)
Nov 09 17:17:24 P14s-Arch keymap[442156]: Selecting devices from the following list:
Nov 09 17:17:24 P14s-Arch keymap[442156]: ------------------------------------------------------------------------------
Nov 09 17:17:24 P14s-Arch keymap[442156]: /dev/input/event0 : Power Button
Nov 09 17:17:24 P14s-Arch keymap[442156]: /dev/input/event1 : Lid Switch
Nov 09 17:17:24 P14s-Arch keymap[442156]: /dev/input/event10: HD-Audio Generic Mic
Nov 09 17:17:24 P14s-Arch keymap[442156]: /dev/input/event11: HD-Audio Generic Headphone
Nov 09 17:17:24 P14s-Arch keymap[442156]: /dev/input/event12: SynPS/2 Synaptics TouchPad
Nov 09 17:17:24 P14s-Arch keymap[442156]: /dev/input/event13: TPPS/2 Elan TrackPoint
Nov 09 17:17:24 P14s-Arch keymap[442156]: /dev/input/event14: xremap pid=442156
Nov 09 17:17:24 P14s-Arch keymap[442156]: /dev/input/event16: USB-HID Keyboard System Control
Nov 09 17:17:24 P14s-Arch keymap[442156]: /dev/input/event17: USB-HID Keyboard Consumer Control
Nov 09 17:17:24 P14s-Arch keymap[442156]: /dev/input/event18: USB-HID Keyboard Mouse
Nov 09 17:17:24 P14s-Arch keymap[442156]: /dev/input/event19: USB-HID Keyboard
Nov 09 17:17:24 P14s-Arch keymap[442156]: /dev/input/event2 : Sleep Button
Nov 09 17:17:24 P14s-Arch keymap[442156]: /dev/input/event20: DP-1
Nov 09 17:17:24 P14s-Arch keymap[442156]: /dev/input/event3 : Video Bus
Nov 09 17:17:24 P14s-Arch keymap[442156]: /dev/input/event4 : AT Translated Set 2 keyboard
Nov 09 17:17:24 P14s-Arch keymap[442156]: /dev/input/event5 : ThinkPad Extra Buttons
Nov 09 17:17:24 P14s-Arch keymap[442156]: /dev/input/event6 : PC Speaker
Nov 09 17:17:24 P14s-Arch keymap[442156]: /dev/input/event7 : HD-Audio Generic HDMI/DP,pcm=3
Nov 09 17:17:24 P14s-Arch keymap[442156]: /dev/input/event8 : HD-Audio Generic HDMI/DP,pcm=7
Nov 09 17:17:24 P14s-Arch keymap[442156]: /dev/input/event9 : HD-Audio Generic HDMI/DP,pcm=8
Nov 09 17:17:24 P14s-Arch keymap[442156]: ------------------------------------------------------------------------------
Nov 09 17:17:24 P14s-Arch keymap[442156]: Selected keyboards automatically since --device options weren't specified:
Nov 09 17:17:25 P14s-Arch keymap[442156]: Failed to grab device 'USB-HID Keyboard' at '/dev/input/event19' due to: No such device (os error 19)
Nov 09 17:17:25 P14s-Arch keymap[442156]: ------------------------------------------------------------------------------
Nov 09 17:17:25 P14s-Arch keymap[442156]: /dev/input/event4 : AT Translated Set 2 keyboard
Nov 09 17:17:25 P14s-Arch keymap[442156]: ------------------------------------------------------------------------------

```

**Comment by k0kubun** [comment]
> multiple times per day, xremap stops working and needs restarting. I don't know if the below error is related to it, but its in the logs periodically so maybe it's related to this.

Is this happening even when you aren't plugging/plugging your keyboard? 

I assume you're using either `--watch` or `--watch=device`. For investigation purposes, what happens if you don't use `--watch` option? Does the device still periodically get considered removed, or does that work as long as you don't unplug the keyboard?

**Comment by xpusostomos** [comment]
@k0kubun I'm not unplugging my keyboard. my setup is I have a laptop, plugged into a monitor, and a keyboard is plugged into the monitor, and I guess the monitor acts like a hub to transmit the signals to the laptop. I don't know what happens when the monitor goes to sleep, I presume it still works because you can wake it with the keyboard. I also don't know if Linux sees it as one device or 2 devices, or whether it's possible when it stops working, it stops for both keyboards. I had assumed so. I'm not using --watch, I can't see documentation on what that is. I guess I'll run it with --watch and see what happens.

**Comment by k0kubun** [comment]
sounds good. Let's try `--watch` then.

**Comment by xpusostomos** [comment]
@k0kubun so far so good, it hasn't stopped working. What does --watch do exactly?

**Comment by xpusostomos** [comment]
been running 4 days now, and still no issues. I'm curious what --watch does and why it isn't the default behavior.

**Comment by k0kubun** [comment]
Thanks for trying and reporting it. It creates an inotify on `/dev/input` and reacts to changes under that directory, e.g. device plug/unplugs. 

The cost to watch that directory is not zero, so it seemed natural to keep it disabled by default. However, I use that myself and can't think of any use case to disable it right now. So I'm open to enabling it by default, even making it impossible to disable it to simplify the interface. WDYT?

**Comment by xpusostomos** [comment]
I would have thought the cost of ionotify would be pretty much zero since it's only activated on device changes. I also think the default configuration should be what works reliably, and options, if they exist, for optimisations. Based on what I know, which isn't much, yes I think it should be the default. And I doubt the need to optimise.

**Comment by k0kubun** [comment]
OK. Would you be interested in filing a PR to make it the default?

**Comment by xpusostomos** [comment]
@k0kubun The trouble is... this is written in Rust right, and I haven't done Rust.

**Comment by choucavalier** [comment]
Hello guys. I have the same issue and stumbled upon this thread. No Rust experience on my side either.

I have a systemd user unit file in `~/.config/systemd/user/xremap.service` and I've added `--watch` to the command that runs xremap. Let's see if that works!
